pandas/io/excel.py

```diff
@@ -615,13 +669,22 @@ def _parse_cell(cell_contents, cell_typ):
                 row = []
                 for j, (value, typ) in enumerate(zip(sheet.row_values(i),
                                                      sheet.row_types(i))):
-                    if usecols is not None and j not in should_parse:
-                        should_parse[j] = self._should_parse(j, usecols)
+                    if ((usecols is not None) or (usecols_excel is not None) or
+                            (j not in should_parse)):
+                        should_parse[j] = self._should_parse(j, usecols_excel,
+                                                             usecols)
 
-                    if usecols is None or should_parse[j]:
+                    if (((usecols_excel is None) and (usecols is None)) or
+                            should_parse[j]):
                         row.append(_parse_cell(value, typ))
                 data.append(row)
 
+            # In 'TextParser' class below 'usecols' can not be an int and at
+            # this point 'usecols' has already been used when it is an int.
+            # So, it receives None so the rest of the flow behaves as expected.
+            if isinstance(usecols, int):
```

Comment:
this is very very odd, what are you trying to do here?

Reply:
I can't see how the parser accepts it.
@jreback take a look to check I'm doing something wrong here.

I've commented the following
```if isinstance(usecols, int):
        usecols = None
``` 
and ran the following test
`pytest pandas/tests/io/test_excel.py -k test_usecols_int`

and the result is:

```
____________________ TestXlrdReader.test_usecols_int[.xlsm] ____________________

self = <pandas.tests.io.test_excel.TestXlrdReader object at 0x7ffa4fe5a278>
ext = '.xlsm'

    def test_usecols_int(self, ext):
    
        dfref = self.get_csv_refdf('test1')
        dfref = dfref.reindex(columns=['A', 'B', 'C'])
>       df1 = self.get_exceldf('test1', ext, 'Sheet1', index_col=0, usecols=3)

pandas/tests/io/test_excel.py:114: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/tests/io/test_excel.py:104: in get_exceldf
    return read_excel(pth, *args, **kwds)
pandas/util/_decorators.py:177: in wrapper
    return func(*args, **kwargs)
pandas/util/_decorators.py:177: in wrapper
    return func(*args, **kwargs)
pandas/io/excel.py:369: in read_excel
    **kwds)
pandas/io/excel.py:498: in parse
    **kwds)
pandas/io/excel.py:759: in _parse_excel
    **kwds)
pandas/io/parsers.py:2008: in TextParser
    return TextFileReader(*args, **kwds)
pandas/io/parsers.py:787: in __init__
    self._make_engine(self.engine)
pandas/io/parsers.py:1024: in _make_engine
    self._engine = klass(self.f, **self.options)
pandas/io/parsers.py:2052: in __init__
    self.usecols, _ = _validate_usecols_arg(kwds['usecols'])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

usecols = 3

    def _validate_usecols_arg(usecols):
        """
        Validate the 'usecols' parameter.
    
        Checks whether or not the 'usecols' parameter contains all integers
        (column selection by index), strings (column by name) or is a callable.
        Raises a ValueError if that is not the case.
    
        Parameters
        ----------
        usecols : list-like, callable, or None
            List of columns to use when parsing or a callable that can be used
            to filter a list of table columns.
    
        Returns
        -------
        usecols_tuple : tuple
            A tuple of (verified_usecols, usecols_dtype).
    
            'verified_usecols' is either a set if an array-like is passed in or
            'usecols' if a callable or None is passed in.
    
            'usecols_dtype` is the inferred dtype of 'usecols' if an array-like
            is passed in or None if a callable or None is passed in.
        """
        msg = ("'usecols' must either be list-like of all strings, all unicode, "
               "all integers or a callable.")
        if usecols is not None:
            if callable(usecols):
                return usecols, None
            # GH20529, ensure is iterable container but not string.
            elif not is_list_like(usecols):
>               raise ValueError(msg)
E               ValueError: 'usecols' must either be list-like of all strings, all unicode, all integers or a callable.

pandas/io/parsers.py:1203: ValueError

```

Reply:
@chris-b1 may you help me here, too, please?

Reply:
it must be a list

Reply:
So, just to be sure. @jreback, you want me to remove support to receives and int in `usecols`?
Is that correct?

Reply:
a single int is not allowed
you should not be special casing anything here
TextParser is the ground truth